hadoop/util.sh

```diff
@@ -1,8 +1,71 @@
 #!/bin/bash
 
 source /usr/lib/hustler/bin/qubole-bash-lib.sh
+source /usr/lib/qubole/bootstrap-functions/common/utils.sh
 export PROFILE_FILE=${PROFILE_FILE:-/etc/profile}
 export HADOOP_ETC_DIR=${HADOOP_ETC_DIR:-/usr/lib/hadoop2/etc/hadoop}
+declare -A SVC_USERS=([namenode]=hdfs [timelineserver]=yarn [historyserver]=mapred [resourcemanager]=yarn [datanode]=hdfs)
+
+function start_daemon() {
+  daemon=$1;
+  case "${SVC_USERS[$daemon]}" in
+    yarn)
+      /bin/su -s /bin/bash -c "/usr/lib/hadoop2/sbin/yarn-daemon.sh start $daemon" yarn
+      ;;
+    hdfs)
+      /bin/su -s /bin/bash -c "/usr/lib/hadoop2/sbin/hadoop-daemon.sh start $daemon" hdfs
+      ;;
+    mapred)
+      /bin/su -s /bin/bash -c "HADOOP_LIBEXEC_DIR=/usr/lib/hadoop2/libexec /usr/lib/hadoop2/sbin/mr-jobhistory-daemon.sh start $daemon" mapred
+      ;;
+    *)
+      echo "Invalid daemon $daemon"
+      ;;
+  esac
+}
+
+function stop_daemon() {
+  daemon=$1;
+  case "${SVC_USERS[$daemon]}" in
+    yarn)
+      /bin/su -s /bin/bash -c "/usr/lib/hadoop2/sbin/yarn-daemon.sh stop $daemon" yarn
+      ;;
+    hdfs)
+      /bin/su -s /bin/bash -c "/usr/lib/hadoop2/sbin/hadoop-daemon.sh stop $daemon" hdfs
+      ;;
+    mapred)
+      /bin/su -s /bin/bash -c "HADOOP_LIBEXEC_DIR=/usr/lib/hadoop2/libexec /usr/lib/hadoop2/sbin/mr-jobhistory-daemon.sh stop $daemon" mapred
+      ;;
+    *)
+      echo "Invalid daemon $daemon"
+      ;;
+  esac
+}
+
+function restart_services() {
+  svcs=("$@")
+  running_svcs=($(get_running_services "${svcs[@]}"))
+  for s in "${running_svcs[@]}"; do
+    monit unmonitor "$s"
+  done
+
+  for s in "${running_svcs[@]}"; do
+    stop_daemon "$s"
+  done
+
+  last=${#running_svcs[@]}
+
+  # Restart services in reverse order of how
+  # they were stopped
+  for (( i=0; i <last; i++ )); do
+    start_daemon "${running_svcs[~i]}"
```

Comment:
I think it's more readable to loop from $last-1 down to 0, instead of using bitwise negated i as the index. But we can leave it like this if it's idiomatic Bash

Reply:
It's pretty standard, but I'm ok with changing it if you feel it's confusing.